### ISSUE TITLE
[Feature]: Add a 3T flagship preset with deep reasoning and frontier-class multimodal quality

### ISSUE THREAD

**Issue by mantreshkhurana** [Spyxpo/framerai#190]
### Preflight checklist

- [x] I searched existing issues and did not find a duplicate.

### Target component

Model (Python)

### Problem statement

The largest preset today is `framer-2t-a49b`: 1.96T text parameters, 49.40B active per
token, plus 39.32B of multimodal towers, for 2.00T in total. Two things limit what a
checkpoint at that size could deliver.

The first is shape. The multimodal towers are 39.32B out of 2.00T. The vision encoder is
12.89B, the image diffusion decoder 10.23B, video 7.33B and the audio stack under 9B, so
the flagship is close to a very large text model with modest towers attached rather than a
system whose image, video and audio paths were scaled with the backbone.

The second is inference behaviour. There is no reasoning mode anywhere in the repository.
`model/generate.py` samples a single forward pass per token with no thinking segment, no
verification step and no way to spend more compute on a harder question. `build.py` has no
reasoning-related flag, the tokenizer has no thinking tokens, and neither the backend nor
the website has a concept of a reasoning budget. A larger model alone does not add
deliberation; the sampling loop has to be able to use it.

TODOS.md already records the honest version of this: parameter count is the ceiling, not
the quality. This issue is about raising the ceiling and building the parts that convert it
into output.

### Proposed solution

Three deliverables, tracked here as one epic.

**1. A `framer-3t-a64b` preset that scales every modality.**

Extending `framer-2t-a49b` rather than redesigning it. Measured with
`model.utils.helpers.estimate_params`, the following shape produces **2.93T text
parameters, 63.98B active per token, plus 76.73B of multimodal towers, for 3.01T total**:

- Backbone unchanged at `d_model=10240`, `n_layers=84`, `max_seq_len=16384`.
- `n_experts` 384 to 512, `expert_d_ff` 2048 to 2304, `n_experts_per_tok` 4 to 6, one
  shared expert retained. 512 is 4 x 128, so it still divides evenly across 8, 16, 32, 64
  and 128-way expert-parallel meshes, which is the reason the current preset chose 384.
  Raising top-k from 4 to 6 is deliberate: active parameters per token move from 49.40B to
  63.98B, and per-token compute is what reasoning-heavy decoding consumes.
- Towers roughly doubled: vision `d_model` 4096 to 5120 at 72 layers (22.66B), audio
  encoder 3072 to 4096 at 56 layers (11.29B), image DiT 3072 to 4096 at 40 layers (21.57B),
  video DiT 2304 to 3072 at 40 layers (15.42B), audio LM 2560 to 3072 at 32 layers.

The configuration passes `FramerConfig.validate()` and can be sized on a laptop without
allocating anything, as the existing presets can. Weights are 6.01 TB in bf16 and 48.1 TB
with AdamW optimiser state, so this is a cluster-only definition; that is stated in the
preset docstring the same way `framer-2t-a49b` states it.

**2. A deep-reasoning path.**

- A reasoning segment in the chat template with dedicated special tokens, so a thinking
  span is a first-class part of the format rather than prose the model happens to emit.
  This depends on the chat template landing in #159.
- A reasoning budget in `model/generate.py`: a token allowance for the thinking span,
  separated from the answer budget, with the thinking span excluded from what the backend
  returns by default.
- A reasoning-effort control (off / low / medium / high) plumbed through `build.py`,
  `model/serve.py`, the backend generation routes and the website settings panel, reusing
  the sampling-control plumbing already added there.
- Self-consistency and a verification pass as opt-in test-time compute strategies, so the
  extra compute has somewhere to go beyond a longer single sample.
- Reuse rather than duplicate the cognition layer: `model/cognition/` already holds working
  memory, recall and an affect model that modulates decoding. The reasoning path should
  consume that trace where it is available instead of introducing a parallel scratchpad.

**3. Generation and understanding quality at the level the architecture allows.**

The architecture roadmap in TODOS.md already names the remaining gaps, and this epic is
what they add up to:

- Image: train the VAE and re-measure `scale_factor`, add adversarial and perceptual
  reconstruction losses, and run the caption enrichment pass. Prompt adherence is dominated
  by caption quality, not by decoder size.
- Video: train the video VAE and re-measure its `scale_factor`, and exercise streaming
  decode, which the causal VAE makes possible but nothing calls.
- Audio: wire the CTC head into the training loop (#158), add adversarial and multi-scale
  spectral losses for the codec, and add prosody conditioning beyond speaker identity.
- Understanding: wire the contrastive trainer into `build.py` as `--mode pretrain-vision`,
  and extend interleaved token placement to the generation path.

### Acceptance criteria

- `framer-3t-a64b` is registered, `--list-presets` and `--estimate` report it, and the
  reported total is at or above 3.0T with active parameters per token reported separately.
- The preset constructs on the meta-device path, and expert sharding is verified for 8, 16,
  32, 64 and 128-way meshes by the placement arithmetic already covered by tests.
- Reasoning effort is settable end to end, from `build.py` and the website through to the
  worker, and the thinking span is separable from the answer in the API response.
- Every claim of improved quality is backed by a number from `model/eval/`. The harness
  already covers text perplexity and bits per byte, image Frechet distance and contrastive
  alignment, audio SI-SDR, mel distance, WER, CER and speaker similarity, and video FVD and
  temporal consistency, with standard benchmark adapters for external comparison. A
  before-and-after table on the same checkpoint is the evidence this issue closes on.

### Dependencies

- #157 grouped GEMM for expert dispatch. At 512 experts across 80 MoE layers the current
  per-expert `index_add_` loop is 40,960 Python-level expert calls per forward, which makes
  this preset untrainable in practice.
- #159 instruction and preference post-training. The chat template it introduces is where
  the reasoning segment lives.
- #158 CTC head wired into training, for audio understanding.
- #156 tokenizer merge cap, since the reasoning tokens are added to the vocabulary.

### Alternatives considered

Growing the dense ladder instead of the expert count. Dense scaling raises active
parameters per token in lockstep with total, so a 3T dense model is not servable on any
realistic budget, while the sparse route keeps per-token compute at 63.98B.

Scaling the text backbone only and leaving the towers where they are. That is cheaper and
lands sooner, but it would repeat the current imbalance and would not move image, video or
audio quality at all.

Prompt-level reasoning, that is, asking the model to think step by step at inference time.
It needs no training work, but without a trained format there is nothing to separate
thinking from the answer, no budget to control and nothing to verify against.

### Additional context

Nothing here removes the external constraints. Training compute and licensed data are a
separate problem from architecture, and the preset is a definition that can be sized and
sharded long before it can be trained. The value of landing it now is that the scaling
arithmetic, the expert-parallel placement and the reasoning format all get exercised
against the shape they will eventually have to hold.

Tracked in TODOS.md under "Architecture roadmap - reaching frontier-class output".

**Comment by mantreshkhurana** [comment]
Sequencing, so this does not get picked up in the wrong order.

The preset itself is the cheapest part of the epic. It is a dictionary of config overrides
and a docstring, it can be validated and sized without a GPU, and it is a day of work. It
should not be done first, because merging a preset nobody can train adds a number to
`--list-presets` and nothing else.

The order that produces something measurable at each step:

1. **#157, grouped GEMM.** At 512 experts across 80 MoE layers, the per-expert
   `index_add_` dispatch is 40,960 Python-level expert calls per forward pass. Until this
   is replaced, the preset is a definition that cannot be trained even on a cluster.
2. **#159, chat template and post-training.** The reasoning segment is a set of special
   tokens and a format contract. It has to live in the same template the tokenizer applies
   at training and inference time, or the serving path drifts from the trained format.
3. **The reasoning path in `model/generate.py`**, then the effort control through
   `model/serve.py`, the backend routes and the website settings panel. Testable at
   `framer-tiny` scale, and it should be tested there, since none of it depends on size.
4. **The quality items** in image, video, audio and understanding. These are independent of
   each other and can proceed in parallel; each has a metric already in `model/eval/`.
5. **The preset**, last, with the estimator output and the expert-parallel placement check
   in the pull request description.

Steps 3 and 4 are worth stating plainly: they are the ones that determine output quality.
Step 5 raises the ceiling. Landing 5 without 3 and 4 would produce a larger model that
behaves exactly like the current one.

**Comment by mantreshkhurana** [comment]
Parameter budget for the proposed shape, measured with
`model.utils.helpers.estimate_params` rather than estimated by hand. Nothing was
instantiated to produce these numbers.

```
framer-2t-a49b:  1.96T text, 49.40B active/token, 39.32B multimodal =  2.00T total
framer-3t-a64b:  2.93T text, 63.98B active/token, 76.73B multimodal =  3.01T total
```

Multimodal breakdown for the proposed preset, in billions:

```
vision_encoder    22.66      image_diffusion   21.57
audio_encoder     11.29      video_diffusion   15.42
audio_codec        5.04      image_vae          0.13
video_vae          0.24      audio_vocoder      0.08
                                       total  76.73
```

Storage and training footprint: 6.01 TB of weights in bf16, and 48.1 TB once AdamW
optimiser state is included. That is the whole reason this stays a cluster-only preset, and
it is the same caveat `framer-2t-a49b` already carries in its docstring.

Two shape notes:

- 512 experts is 4 x 128, which keeps even division across 8, 16, 32, 64 and 128-way
  expert-parallel meshes. The current preset picked 384 for the same reason, and dropping
  that property would break `model/training/expert_parallel.py` placement on the larger
  meshes.
- Top-k moving from 4 to 6 is the only change that raises per-token cost, from 49.40B to
  63.98B active. It is proposed on purpose. Reasoning decoding spends its budget on tokens,
  and the width of the path each token takes is what it has to work with.

If the target is a round 3.0T rather than the shape above, `n_experts=640` with
`expert_d_ff=1920` and top-k 8 reaches 3.09T at 66.91B active. It is a worse trade: more
routing overhead for the same parameter count.
